rust/src/array.rs

```diff
@@ -15,297 +15,421 @@
 // specific language governing permissions and limitations
 // under the License.
 
+///! Array types
+use std::any::Any;
 use std::convert::From;
-use std::iter::Iterator;
+use std::ops::Add;
 use std::rc::Rc;
 use std::str;
 use std::string::String;
 
 use super::bitmap::Bitmap;
-use super::buffer::Buffer;
-use super::list::List;
-
-pub enum ArrayData {
-    Boolean(Buffer<bool>),
-    Float32(Buffer<f32>),
-    Float64(Buffer<f64>),
-    Int8(Buffer<i8>),
-    Int16(Buffer<i16>),
-    Int32(Buffer<i32>),
-    Int64(Buffer<i64>),
-    UInt8(Buffer<u8>),
-    UInt16(Buffer<u16>),
-    UInt32(Buffer<u32>),
-    UInt64(Buffer<u64>),
-    Utf8(List<u8>),
-    Struct(Vec<Rc<Array>>),
-}
+use super::buffer::*;
+use super::builder::*;
+use super::datatypes::*;
+use super::list::*;
+use super::list_builder::*;
 
-macro_rules! arraydata_from_primitive {
-    ($DT:ty, $AT:ident) => {
-        impl From<Vec<$DT>> for ArrayData {
-            fn from(v: Vec<$DT>) -> Self {
-                ArrayData::$AT(Buffer::from(v))
-            }
-        }
-        impl From<Buffer<$DT>> for ArrayData {
-            fn from(v: Buffer<$DT>) -> Self {
-                ArrayData::$AT(v)
-            }
-        }
-    };
+/// Trait for dealing with different types of Array at runtime when the type of the
+/// array is not known in advance
+pub trait Array {
+    /// Returns the length of the array (number of items in the array)
+    fn len(&self) -> usize;
+    /// Returns the number of null values in the array
+    fn null_count(&self) -> usize;
+    /// Optional validity bitmap (can be None if there are no null values)
+    fn validity_bitmap(&self) -> &Option<Bitmap>;
+    /// Return the array as Any so that it can be downcast to a specific implementation
+    fn as_any(&self) -> &Any;
 }
 
-arraydata_from_primitive!(bool, Boolean);
-arraydata_from_primitive!(f32, Float32);
-arraydata_from_primitive!(f64, Float64);
-arraydata_from_primitive!(i8, Int8);
-arraydata_from_primitive!(i16, Int16);
-arraydata_from_primitive!(i32, Int32);
-arraydata_from_primitive!(i64, Int64);
-arraydata_from_primitive!(u8, UInt8);
-arraydata_from_primitive!(u16, UInt16);
-arraydata_from_primitive!(u32, UInt32);
-arraydata_from_primitive!(u64, UInt64);
-
-pub struct Array {
-    /// number of elements in the array
-    len: i32,
-    /// number of null elements in the array
-    null_count: i32,
-    /// If null_count is greater than zero then the validity_bitmap will be Some(Bitmap)
+/// Array of List<T>
+pub struct ListArray<T: ArrowPrimitiveType> {
```

Comment:
Seems there's no constructor to initialize `null_count` and `validity_bitmap` (`validity_bitmap` is always `None` and `null_count` is always 0). Should we add one?

Reply:
As far as I remember, `null_count` is a cached count extracted from `validity_bitmap`. I'm not sure we even need this as a member here, since this could also be handled by the `BitMap` type, and a member method `ListArray::null_count(&self) -> usize` could return `0` if `validity_bitmap` is `None`. For now, this may be OK and this could be done in another PR.

Reply:
Buffer<T> does not support strings. It only supports fixed-width types (bool, ints, floats). Strings are represented by List<u8>. I believe the Arrow spec states that all languages must support the same endianness for these types to ensure compatibility. Arrow doesn't have a bit-packed boolean type AFAIK.

Regarding the null count / constructor questions, we generally need to use the builders to build Buffer/List arrays that contain null values. There is more we can do I'm sure but the focus of this PR is for changing the underlying data structures from enum to trait/generics. We didn't have great null handling before either and I think we should handle that in a separate PR once we've taken care of this refactor if there are no objections.

Reply:
> Arrow doesn't have a bit-packed boolean type AFAIK.

No, the boolean type is always bit-packed according to the spec and implemented as such in Java and C++

Reply:
OK, but this isn't implemented right now and we could do this in another PR. So this boolean types probably won't be primitive types. (that's one possible solution)